src/topology/noetherian_space.lean

```diff
@@ -0,0 +1,217 @@
+
+namespace topological_space
+
+/-- Type class for noetherian spaces. It is defined to be spaces whose open sets satisfies ACC. -/
+@[mk_iff]
+class noetherian_space : Prop :=
+(well_founded : well_founded ((>) : opens α → opens α → Prop))
+
+lemma noetherian_space_iff_opens :
+  noetherian_space α ↔ ∀ s : opens α, is_compact (s : set α) :=
+begin
+  rw [noetherian_space_iff, complete_lattice.well_founded_iff_is_Sup_finite_compact,
+    complete_lattice.is_Sup_finite_compact_iff_all_elements_compact],
+  exact forall_congr opens.is_compact_element_iff,
+end
+
+@[priority 100]
+instance noetherian_space.compact_space [h : noetherian_space α] : compact_space α :=
+is_compact_univ_iff.mp ((noetherian_space_iff_opens α).mp h ⊤)
+
+variable {α}
+
+instance noetherian_space.set [h : noetherian_space α] (s : set α) : noetherian_space s :=
+begin
+  rw noetherian_space_iff,
+  apply well_founded.well_founded_iff_has_max'.2,
+  intros p hp,
+  obtain ⟨⟨_, u, hu, rfl⟩, hu'⟩ := hp,
+  obtain ⟨U, hU, hU'⟩ := well_founded.well_founded_iff_has_max'.1 h.1
+    (((opens.comap ⟨_, continuous_subtype_coe⟩)) ⁻¹' p) ⟨⟨u, hu⟩, hu'⟩,
+  refine ⟨opens.comap ⟨_, continuous_subtype_coe⟩ U, hU, _⟩,
+  rintros ⟨_, x, hx, rfl⟩ hx' hx'',
+  refine le_antisymm (set.preimage_mono (_ : (⟨x, hx⟩ : opens α) ≤ U)) hx'',
+  refine sup_eq_right.mp (hU' (⟨x, hx⟩ ⊔ U) _ le_sup_right),
+  dsimp [set.preimage],
+  rw map_sup,
+  convert hx',
+  exact sup_eq_left.mpr hx''
+end
+
+variable (α)
+
+example (α : Type*) : set α ≃o (set α)ᵒᵈ := by refine order_iso.compl (set α)
+
+lemma noetherian_space_tfae :
+  tfae [noetherian_space α,
+    well_founded (λ s t : closeds α, s < t),
+    ∀ s : set α, is_compact s,
+    ∀ s : opens α, is_compact (s : set α)] :=
+begin
+  tfae_have : 1 ↔ 2,
+  { refine (noetherian_space_iff _).trans (surjective.well_founded_iff opens.compl_bijective.2 _),
+    exact λ s t, (order_iso.compl (set α)).lt_iff_lt.symm },
+  tfae_have : 1 ↔ 4,
+  { exact noetherian_space_iff_opens α },
+  tfae_have : 1 → 3,
+  { intros H s, rw is_compact_iff_compact_space, resetI, apply_instance },
+  tfae_have : 3 → 4,
+  { exact λ H s, H s },
+  tfae_finish
+end
+
+variables {α β}
+
+lemma noetherian_space.is_compact [h : noetherian_space α] (s : set α) : is_compact s :=
+let H := (noetherian_space_tfae α).out 0 2 in H.mp h s
+
+lemma noetherian_space_of_surjective [noetherian_space α] (f : α → β)
+  (hf : continuous f) (hf' : function.surjective f) : noetherian_space β :=
+begin
+  rw noetherian_space_iff_opens,
+  intro s,
+  obtain ⟨t, e⟩ := set.image_surjective.mpr hf' s,
+  exact e ▸ (noetherian_space.is_compact t).image hf,
+end
+
+lemma noetherian_space.range [noetherian_space α] (f : α → β) (hf : continuous f) :
+  noetherian_space (set.range f) :=
+noetherian_space_of_surjective (set.cod_restrict f _ set.mem_range_self) (by continuity)
+  (λ ⟨a, b, h⟩, ⟨b, subtype.ext h⟩)
+
+lemma noetherian_space_set_iff (s : set α) :
+  noetherian_space s ↔ ∀ t ⊆ s, is_compact t :=
```

Comment:
If we were to mimick `is_compact` and `compact_space`, we could take this to be the definition for a set to be noetherian, and define `noetherian_space` to mean that `set.univ` is noetherian. Given that we are not doing that, I think we need to add a lemma analogous to [is_compact_univ_iff](https://leanprover-community.github.io/mathlib_docs/topology/subset_properties.html#is_compact_univ_iff), which would allow you to prove the underlying space of any noetherian scheme is noetherian from the fact that the spectra of noetherian rings are noetherian and `noetherian_space.Union` below.

Reply:
I'm not planning to add a separate predicate for noetherian subsets since I have hardly ever heard about a noetherian set. Plus you could always coerce it into a Type.

I don't really know what you mean by "a lemma analogous to is_compact_univ_iff"?
I believe `noetherian_space.Union` and `noetherian_space.range` alone suffices to prove that the underlying noetherian schemes are noetherian.

Reply:
I think what you need is `noetherian_space_of_surjective` applied to [homeomorph.set.univ](https://leanprover-community.github.io/mathlib_docs/topology/homeomorph.html#homeomorph.set.univ). When combined they give you `noetherian_space α ↔ noetherian_space (set.univ : set α)`, which is what I meant by "a lemma analogous to is_compact_univ_iff".

Reply:
Done!